driver/src/price_finding/naive_solver.rs

```diff
@@ -0,0 +1,74 @@
+use crate::models;
+
+use super::error::PriceFindingError;
+use super::price_finder_interface::Solution;
+use web3::types::U256;
+use crate::models::Order;
+
+impl Order {
+    fn matches(&self, other: &Order) -> bool {
+        self.opposite_tokens(other) && self.have_price_overlap(other)
+    }
+
+    fn opposite_tokens(&self, other: &Order) -> bool {
+        self.buy_token == other.sell_token && self.sell_token == other.buy_token
+    }
+
+    fn have_price_overlap(&self, other: &Order) -> bool {
+        self.buy_amount * other.buy_amount <= other.sell_amount * self.sell_amount
+    }
+}
+
+pub fn solve(orders: &Vec<models::Order>, num_tokens: u8) -> Result<Solution, PriceFindingError> {
+    // TODO - use mapping here
+    let mut prices: Vec<u128> = vec![0; num_tokens as usize];
+    let mut sell_amount: Vec<u128> = vec![0; orders.len()];
+    let mut buy_amount: Vec<u128> = vec![0; orders.len()];
+
+    for (i, x) in orders.iter().enumerate() {
+        for j in i + 1..orders.len() {
+            let y = &orders[j];
+            if x.matches(y) {
+                if x.buy_amount <= y.sell_amount && x.sell_amount <= y.buy_amount {
+                    // Type I-A (x <= y)
+                    prices[x.buy_token - 1] = x.sell_amount;
+                    prices[y.buy_token - 1] = x.buy_amount;
+
+                    sell_amount[i] = x.sell_amount;
+                    sell_amount[j] = x.buy_amount;
+
+                    buy_amount[i] = x.buy_amount;
+                    buy_amount[j] = x.sell_amount;
+                } else if x.buy_amount >= y.sell_amount && x.sell_amount >= y.buy_amount {
+                    // Type I-B (y <= x)
+                    prices[x.buy_token - 1] = y.sell_amount;
+                    prices[y.buy_token - 1] = y.buy_amount;
+
+                    sell_amount[i] = y.sell_amount;
+                    sell_amount[j] = y.buy_amount;
+
+                    buy_amount[i] = y.buy_amount;
+                    buy_amount[j] = y.sell_amount;
+                } else {
+                    // Type II
+                    prices[x.buy_token - 1] = y.sell_amount;
+                    prices[y.buy_token - 1] = x.sell_amount;
+
+                    sell_amount[i] = x.sell_amount;
+                    sell_amount[j] = y.sell_amount;
+
+                    buy_amount[i] = y.sell_amount;
+                    buy_amount[j] = x.sell_amount;
+                }
+                break;
```

Comment:
As with the python code, I find it somewhat counterintuitive that we set everything up to be looped but in reality we only want to find the first match and then return immediately. Maybe we can prepare the solution inside the loop and return there, and otherwise just return `trivialSolution()`

Reply:
I really thought I had already replaces this break with a return statement...